lib/tbot/service_spiffe_workload_api.go

```diff
@@ -0,0 +1,489 @@
+	// fails, we need to close the client.
+	defer func() {
+		if err != nil {
+			client.Close()
+		}
+	}()
+
+	if err := s.fetchBundle(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	authPing, err := client.Ping(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	s.trustDomain = authPing.ClusterName
+
+	return nil
+}
+
+func createListener(addr string) (net.Listener, error) {
+	parsed, err := url.Parse(addr)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing %q", addr)
+	}
+
+	switch parsed.Scheme {
+	case "tcp":
+		return net.Listen("tcp", parsed.Host)
+	case "unix":
+		return net.Listen("unix", parsed.Path)
+	default:
+		return nil, trace.BadParameter("unsupported scheme %q", parsed.Scheme)
+	}
+}
+
+func (s *SPIFFEWorkloadAPIService) Run(ctx context.Context) error {
+	ctx, span := tracer.Start(ctx, "SPIFFEWorkloadAPIService/Run")
+	defer span.End()
+	if !experiment.Enabled() {
+		return trace.BadParameter("workload identity has not been enabled")
+	}
+
+	s.log.Info("Completing pre-run initialization")
+	if err := s.setup(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	defer s.client.Close()
+	s.log.Info("Completed pre-run initialization")
+
+	srvMetrics := metrics.CreateGRPCServerMetrics(
+		true, prometheus.Labels{
+			teleport.TagServer: "tbot-spiffe-workload-api",
+		},
+	)
+	if err := metrics.RegisterPrometheusCollectors(srvMetrics); err != nil {
+		return trace.Wrap(err)
+	}
+	srv := grpc.NewServer(
+		grpc.Creds(
+			// SPEC (SPIFFE_Workload_endpoint) 3. Transport:
+			// - Transport Layer Security MUST NOT be required
+			// TODO(noah): We should optionally provide TLS support here down
+			// the road.
+			insecure.NewCredentials(),
+		),
+		grpc.ChainUnaryInterceptor(
+			recovery.UnaryServerInterceptor(),
+			srvMetrics.UnaryServerInterceptor(),
+		),
+		grpc.ChainStreamInterceptor(
+			recovery.StreamServerInterceptor(),
+			srvMetrics.StreamServerInterceptor(),
+		),
+		grpc.StatsHandler(otelgrpc.NewServerHandler()),
+		grpc.MaxConcurrentStreams(defaults.GRPCMaxConcurrentStreams),
+	)
+	workloadpb.RegisterSpiffeWorkloadAPIServer(srv, s)
+
+	lis, err := createListener(s.cfg.Listen)
+	if err != nil {
+		return trace.Wrap(err, "creating listener")
+	}
+	defer func() {
+		if err := lis.Close(); err != nil {
+			s.log.WithError(err).Error("Encountered error closing listener")
+		}
+	}()
+	s.log.WithField("addr", lis.Addr().String()).Info("Listener opened for Workload API endpoint")
+	if lis.Addr().Network() == "tcp" {
+		s.log.Warn(
+			"Workload API endpoint listening on a TCP port. Ensure that only intended hosts can reach this port!",
+		)
+	}
+
+	// Set off the long running tasks in an errgroup
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		// Start the gRPC server
+		return srv.Serve(lis)
+	})
+	eg.Go(func() error {
+		// Shutdown the server when the context is canceled
+		<-egCtx.Done()
+		s.log.Debug("Shutting down Workload API endpoint")
+		srv.Stop()
```

Comment:
Consider whether we want the hard `Stop` vs `GracefulStop` here

Reply:
Good catch, I had GracefulStop originally but the long-lived nature of the Workload API streaming RPCs means that GracefulStop will just hang as these connections don't close.